src/audio/kpb.c

```diff
@@ -0,0 +1,541 @@
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *   * Neither the name of the Intel Corporation nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Marcin Rajwa <marcin.rajwa@linux.intel.com>
+ *
+ */
+
+#include <stdint.h>
+#include <uapi/ipc/topology.h>
+#include <sof/ipc.h>
+#include <sof/audio/component.h>
+#include <sof/audio/kpb.h>
+#include <sof/list.h>
+#include <sof/audio/buffer.h>
+
+static int kpb_register_client(struct comp_dev *dev, struct kpb_client *cli);
+static int kpb_begin_drainning(struct comp_dev *dev, uint8_t client_id);
+static void kpb_buffer_data(struct comp_data *kpb, struct comp_buffer *source);
+
+/**
+ * kpb_new() - create a key phrase buffer component.
+ * @arg1: generic ipc component pointer.
+ *
+ * Return: A pointer to newly created KPB component.
+ */
+static struct comp_dev *kpb_new(struct sof_ipc_comp *comp)
+{
+
+	struct comp_dev *dev;
+	struct sof_ipc_comp_kpb *kpb;
+	struct sof_ipc_comp_kpb *ipc_kpb = (struct sof_ipc_comp_kpb *)comp;
+	struct comp_data *cd;
+
+	trace_kpb("kpb_new()");
+
+	if (IPC_IS_SIZE_INVALID(ipc_kpb->config)) {
+		IPC_SIZE_ERROR_TRACE(TRACE_CLASS_KPB, ipc_kpb->config);
+		return NULL;
+	}
+
+	/* Validate input parameters */
+	if (ipc_kpb->channels > MAX_SUPPORETED_CHANNELS) {
+		trace_kpb_error("kpb_new() error: "
+		"nr of channels exceeded the limit");
+	return NULL;
+	}
+
+	if (ipc_kpb->history_depth > MAX_BUFFER_SIZE) {
+		trace_kpb_error("kpb_new() error: "
+		"history depth exceeded the limit");
+	return NULL;
+	}
+
+	if (ipc_kpb->sampling_freq != MAX_SAMPLNG_FREQUENCY) {
```

Comment:
If there's single supported sampling rate should this macro be without "MAX"?

Reply:
Applied.